interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,392 @@
+		if !strutil.ListContains(permissions, perm) {
+			newPermissions = append(newPermissions, perm)
+		}
+	}
+	if len(newPermissions) != len(pc.Permissions) {
+		pc.Permissions = newPermissions
+		return true
+	}
+	return false
+}
+
+// userPromptDB maps prompt IDs to prompts for a single user.
+type userPromptDB struct {
+	ByID map[string]*Prompt
+}
+
+// PromptDB stores outstanding prompts in memory and ensures that new prompts
+// are created with a unique ID.
+type PromptDB struct {
+	// Mmap the max ID file to avoid unnecessary syscalls
+	maxIDMmap []byte
+	// The per-user DB is protected by a RWMutex
+	mutex   sync.RWMutex
+	perUser map[uint32]*userPromptDB
+	// Function to issue a notice for a change in a prompt
+	notifyPrompt func(userID uint32, promptID string, data map[string]string) error
+}
+
+const (
+	maxIDFileSize                int = 8
+	maxOutstandingPromptsPerUser int = 1000
+)
+
+// New creates and returns a new prompt database.
+//
+// The given notifyPrompt closure will be called when a prompt is added,
+// merged, modified, or resolved.
+func New(notifyPrompt func(userID uint32, promptID string, data map[string]string) error) (*PromptDB, error) {
+	pdb := PromptDB{
+		perUser:      make(map[uint32]*userPromptDB),
+		notifyPrompt: notifyPrompt,
+	}
+	maxIDFilepath := filepath.Join(dirs.SnapRunDir, "request-prompt-max-id")
+	maxIDFile, err := os.OpenFile(maxIDFilepath, os.O_RDWR|os.O_CREATE, 0600)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open max ID file: %w", err)
+	}
+	// The file/FD can be safely closed once the mmap is created.
+	defer maxIDFile.Close()
+	fileInfo, err := maxIDFile.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("cannot stat max ID file: %w", err)
+	}
+	if fileInfo.Size() != int64(maxIDFileSize) {
+		if fileInfo.Size() != 0 {
+			// Max ID file malformed, best to reset it
+			logger.Debugf("requestprompts: max ID file malformed; re-initializing")
+		}
+		if err = initializeMaxIDFile(maxIDFile); err != nil {
+			return nil, err
+		}
+	}
+	pdb.maxIDMmap, err = unix.Mmap(int(maxIDFile.Fd()), 0, maxIDFileSize, unix.PROT_READ|unix.PROT_WRITE, unix.MAP_SHARED)
+	if err != nil {
+		return nil, fmt.Errorf("cannot mmap max ID file: %w", err)
+	}
+	return &pdb, nil
+}
+
+// initializeMaxIDFile truncates the given file to 8 bytes of zeros.
+func initializeMaxIDFile(maxIDFile *os.File) (err error) {
+	if err = maxIDFile.Truncate(8); err != nil {
+		return fmt.Errorf("cannot truncate max ID file: %w", err)
+	}
+	initial := []byte{0, 0, 0, 0, 0, 0, 0, 0}
+	if _, err = maxIDFile.WriteAt(initial, 0); err != nil {
+		return fmt.Errorf("cannot initialize max ID file: %w", err)
+	}
+	return nil
+}
+
+// nextID increments the internal monotonic maxID integer and returns the
+// corresponding ID string.
+func (pdb *PromptDB) nextID() string {
+	// Byte order will be consistent, and want an atomic increment.
+	id := atomic.AddUint64((*uint64)(unsafe.Pointer(&pdb.maxIDMmap[0])), 1)
```

Comment:
I would like to, but there are two problems:
1. `MmapPtr` is only available in the latest version of `x/sys/unix` v0.22.0, while we're on v0.7.0 (and if we create a temporary `unixMmapPtr` function which uses `unix.Mmap()`, we need to hold onto the returned `[]byte` so that it is not GC'd and finalized, causing an internal `Munmap()`)
2. If we use `MmapPtr()`, there's no automatic `Munmap()` if the prompt DB is abandoned without explicitly calling `Close()`, `Munmap` is never called. This is a problem because we don't currently need to call `Close()` unless we care about cleaning up outstanding prompts, which we don't in e.g. unit tests. In production, snapd should be calling `Close()` when shutting down, and if snapd crashes, the kernel should clean up the memory mapping, so perhaps this isn't a huge concern -- we just need to clean up the unit tests.